src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs

```diff
@@ -1003,6 +1010,42 @@ public bool ProcessTransaction(Transaction transaction, int? blockHeight = null,
 
             lock (this.lockObject)
             {
+                // Do a pre-scan of the transaction inputs to see if they're used in other transactions already
+                foreach (TxIn input in transaction.Inputs)
+                {
+                    // See if this input is being used by another transaction.
+                    if (this.inputLookup.TryGetValue(input.PrevOut, out WalletIndexData indexData))
+                    {
+                        if (indexData.TransactionData.Id == hash)
+                            continue; // It's the same transaction, ignore.
+
+                        // Incoming transaction confirmed, existing transaction unconfirmed, delete existing
+                        if ((block != null) && (indexData.TransactionData.BlockHash == null))
+                        {
+                            this.RemoveTransactionsByIdsLocked(indexData.Wallet.Name, new[] { indexData.TransactionData.Id });
+                        }
+
+                        // Incoming transaction unconfirmed, existing transaction unconfirmed, delete existing (so that only 1 is left - either could be the 'correct' version)
```

Comment:
So here we effectively take the "superseding" transaction? I.e. we are replacing the one in lookup dictionary?

Reply:
Yes, the 4 scenarios determine which of the two possibilities (incoming or existing) are the 'better' one and acts accordingly. And yes, the index dictionary gets rebuilt as part of the removal process.